protocol.go

```diff
@@ -23,47 +23,76 @@ import (
 	"golang.org/x/net/context"
 	"google.golang.org/grpc/peer"
 	"net"
+	"sync"
 )
 
 type Protocol interface {
-	Client(Info, context.Context, string, net.Conn) (net.Conn, error)
-	Server(Info, net.Conn) (net.Conn, error)
+	Client(*Info, context.Context, string, net.Conn) (net.Conn, error)
+	Server(*Info, net.Conn) (net.Conn, error)
 }
 
-func InfoFromPeer(peer *peer.Peer) Info {
+func InfoFromPeer(peer *peer.Peer) *Info {
 	if peer.AuthInfo == nil {
 		return nil
 	}
 
-	return peer.AuthInfo.(Info)
+	return peer.AuthInfo.(*Info)
 }
 
-type Info map[string]interface{}
+// Info is thread safe.
+type Info struct {
+	data map[string]interface{}
+	mu   sync.RWMutex
+}
+
+func NewInfo() *Info {
+	return &Info{
+		data: make(map[string]interface{}),
+	}
+}
 
-func (Info) AuthType() string {
+func (*Info) AuthType() string {
 	return "noise"
 }
 
-func (i Info) Put(key string, val interface{}) {
-	i[key] = val
+func (i *Info) Put(key string, val interface{}) {
+	i.mu.Lock()
+	i.data[key] = val
+	i.mu.Unlock()
```

Comment:
Our style guide says: "Use `defer` wherever possible for releasing mutices or resources"